### PR TITLE
Add `bee-identity` crate

### PR DESCRIPTION
**Pull request by Alex6323**
# Description of change

Introduces the `bee-identity` crate, which encapsulates all things required (Ed25519 keypair, PeerId, etc.)  to set up a node identity.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-identity/src/pem_file.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+fn pem_entry_to_keypair(pem_entry: String) -> Result<Keypair, PemFileError> {
+    let KeypairBytes { mut secret_key, .. } = KeypairBytes::from_pkcs8_pem(&pem_entry).or(Err(PemFileError::Parse))?;
+    let secret = SecretKey::from_bytes(&mut secret_key).or(Err(PemFileError::DecodeKeypair))?;
+    Ok(secret.into())
+}
+
+fn keypair_to_pem_entry(keypair: &Keypair) -> Result<String, PemFileError> {
+    // Safety: We know that this has the correct size.
```
Comment (grtlr) [inline bee-identity/src/pem_file.rs]: We might not need that comment anymore.

File: bee-identity/src/pem_file.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+//! Contains types and methods to persist identities using the PEM file format.
+
+use crate::keys::{Ed25519Keypair as Keypair, Ed25519PrivateKey as SecretKey};
+
+use ed25519::KeypairBytes;
+use pkcs8::{DecodePrivateKey, EncodePrivateKey, LineEnding};
+
+use std::{fs, path::Path};
+
+/// TODO
```
Comment (grtlr) [inline bee-identity/src/pem_file.rs]: Maybe allow missing comments here.

File: bee-identity/src/pem_file.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        secret_key,
+        public_key: None,
+    };
+    match keypair_bytes.to_pkcs8_pem(LineEnding::default()) {
+        Ok(zeroize_string) => Ok(zeroize_string.to_string()),
+        Err(_) => Err(PemFileError::DecodeKeypair),
+    }
+}
+
+/// TODO
```
Comment (grtlr) [inline bee-identity/src/pem_file.rs]: Todo

File: bee-identity/src/pem_file.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+pub fn read_keypair_from_pem_file<P: AsRef<Path>>(path: P) -> Result<Keypair, PemFileError> {
+    match fs::read_to_string(path) {
+        Ok(pem_file) => pem_entry_to_keypair(pem_file),
+        Err(e) => Err(PemFileError::Read(e)),
+    }
+}
+
+/// TODO
```
Comment (grtlr) [inline bee-identity/src/pem_file.rs]: Todo

File: bee-network/bee-autopeering/src/init.rs [hunk 5]
```diff
@@ -41,9 +41,9 @@ const BOOTSTRAP_UPDATE_DELAY: Duration = Duration::from_secs(4 * SECOND);
 /// Initializes the autopeering service.
 pub async fn init<S, I, Q, V>(
     config: AutopeeringConfig,
+    local: Local,
     version: u32,
     network_name: I,
-    local: Local,
```
Comment (grtlr) [inline bee-network/bee-autopeering/src/init.rs:46]: Why was the order changed?

File: bee-network/bee-autopeering/src/local/mod.rs [hunk 6]
```diff
@@ -70,8 +71,14 @@ impl Local {
         }
     }
 
-    /// Creates a local identity from an ED25519 keypair.
-    pub fn from_keypair(keypair: Keypair) -> Result<Self, Error> {
+    /// Creates a local entity from a `LocalId`.
+    pub fn from_local_id(local_id: Identity) -> Self {
+        let keypair = local_id.into();
+        Self::from_ed25519_keypair(keypair).expect("faulty local id")
```
Comment (grtlr) [inline bee-network/bee-autopeering/src/local/mod.rs]: Is it OK that this can panic?

File: bee-network/bee-autopeering/src/peer/peer_id.rs [hunk 7]
```diff
@@ -115,6 +117,14 @@ impl AsRef<[u8]> for PeerId {
     }
 }
 
+impl From<Identity> for PeerId {
+    fn from(identity: Identity) -> Self {
+        Self::from_public_key(
+            PublicKey::try_from_bytes(identity.keypair().public().encode()).expect("public key from bytes"),
```
Comment (grtlr) [inline bee-network/bee-autopeering/src/peer/peer_id.rs:123]: Same here

File: bee-network/bee-gossip/src/network/host.rs [hunk 8]
```diff
@@ -226,12 +226,7 @@ async fn dial_peer(swarm: &mut Swarm<SwarmBehaviour>, peer_id: PeerId, peerlist:
         })
         .expect("peer must exist");
 
-    debug!(
-        "Dialing peer: {} ({}) attempt: #{}.",
-        alias,
-        alias!(peer_id),
-        dial_attempt
-    );
+    debug!("Dialing peer: {} ({}) attempt: #{}.", alias, peer_id, dial_attempt);
```
Comment (grtlr) [inline bee-network/bee-gossip/src/network/host.rs]: You absolutely don't have to do this, but I think this is easier to parse:
```suggestion
    debug!("Dialing peer: {alias} ({peer_id}) attempt: #{dial_attempt}.");
```

File: bee-network/bee-gossip/src/tests/common/await_events.rs [hunk 9]
```diff
@@ -61,33 +43,15 @@ pub async fn get_added_peer_id(rx: &mut NetworkEventReceiver) -> PeerId {
...
-                panic!("timed out before receiving `PeerConnected` event");
-            }
-        }
-    }
-}
-
-pub async fn get_gossip_channels(rx: &mut NetworkEventReceiver) -> (GossipReceiver, GossipSender) {
-    let timeout = time::sleep(Duration::from_secs(20));
+pub async fn await_peer_connected_event(rx: &mut NetworkEventReceiver) -> (PeerId, GossipSender, GossipReceiver) {
+    let timeout = time::sleep(Duration::from_secs(1));
```
Comment (grtlr) [inline bee-network/bee-gossip/src/tests/common/await_events.rs]: Just checking: Is the `Duration` change on purpose here?

File: bee-node/bee-node/src/identity.rs [hunk 10]
```diff
@@ -1,21 +1,28 @@
 // Copyright 2020-2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use bee_gossip::{Keypair, SecretKey};
+//! TODO
```
Comment (grtlr) [inline bee-node/bee-node/src/identity.rs]: Todo

File: bee-node/bee-node/src/identity.rs [hunk 11]
```diff
@@ -1,21 +1,28 @@
...
+
+use crate::keypair::{Keypair, SecretKey};
 
 use ed25519::KeypairBytes;
 use pkcs8::{DecodePrivateKey, EncodePrivateKey, LineEnding};
 
 use std::{fs, path::Path};
 
+/// TODO
```
Comment (grtlr) [inline bee-node/bee-node/src/identity.rs]: Todo

File: bee-node/bee-node/src/identity.rs [hunk 12]
```diff
@@ -43,13 +50,15 @@ fn keypair_to_pem_entry(keypair: &Keypair) -> Result<String, PemFileError> {
     }
 }
 
+/// TODO
```
Comment (grtlr) [inline bee-node/bee-node/src/identity.rs]: Todo

File: bee-node/bee-node/src/identity.rs [hunk 13]
```diff
@@ -43,13 +50,15 @@ fn keypair_to_pem_entry(keypair: &Keypair) -> Result<String, PemFileError> {
...
 pub fn read_keypair_from_pem_file<P: AsRef<Path>>(path: P) -> Result<Keypair, PemFileError> {
     match fs::read_to_string(path) {
         Ok(pem_file) => pem_entry_to_keypair(pem_file),
         Err(e) => Err(PemFileError::Read(e)),
     }
 }
 
+/// TODO
```
Comment (grtlr) [inline bee-node/bee-node/src/identity.rs]: Todo

File: bee-node/bee-node/src/util.rs [hunk 14]
```diff
@@ -20,9 +21,9 @@ pub(crate) fn create_id_from_network_name(network_name: &impl AsRef<str>) -> u64
 }
 
 /// Determines name and current version of the node.
-pub(crate) fn create_node_info() -> NodeInfo {
+pub(crate) fn create_node_info(identity: &Identity) -> NodeInfo {
     NodeInfo {
-        name: BEE_NAME.to_owned(),
+        name: identity.alias().to_string(),
```
Comment (grtlr) [inline bee-node/bee-node/src/util.rs]: I think it would be nice to have `Bee` part of the name. Iirc, this name is also returned by the API. Hornet just uses `HORNET` so, I think we should not deviate to much from it. Maybe something like `Bee (<identity>)`?

File: bee-test/src/id.rs [hunk 15]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use bee_gossip::PeerId;
+
+use bee_identity::{Ed25519Keypair, Ed25519PrivateKey};
+
+use std::iter::repeat;
+
+/// TODO
```
Comment (grtlr) [inline bee-test/src/id.rs]: Todo

File: bee-identity/src/identity.rs [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+/// Represents a local identity.
+#[derive(Clone, Debug)]
+pub struct Identity {
+    /// The identity alias.
+    alias: String,
+    /// An Ed25519 keypair.
+    keypair: Ed25519Keypair,
+    /// Whether the identity was newly generated.
+    is_new: bool,
```
Comment (grtlr) [inline bee-identity/src/identity.rs]: I don't know how to do this differently at the moment, but I feel like this is an implementation detail of `bee-node` that is leaking into this crate.

**Review by grtlr — CHANGES_REQUESTED** [review]

File: bee-network/bee-autopeering/src/init.rs (hunk 5 above)
Comment (Alex6323) [reply bee-network/bee-autopeering/src/init.rs:46]: I think for consistency reasons with `bee-gossip`? Do you think the previous order was better?

File: bee-identity/src/identity.rs (hunk 16 above)
Comment (Alex6323) [reply bee-identity/src/identity.rs]: Hmm, I see. I try to come up with a solution.

File: bee-network/bee-autopeering/src/init.rs (hunk 5 above)
Comment (grtlr) [reply bee-network/bee-autopeering/src/init.rs:46]: I like the new order too 😊

File: bee-identity/src/identity.rs (hunk 16 above)
Comment (Alex6323) [reply bee-identity/src/identity.rs]: I moved the migration stuff into `bee-identity` which I found was the cleanest solution. Whether an `identity` is new/fresh is now completely contained.

File: bee-network/bee-gossip/src/network/host.rs (hunk 8 above)
Comment (Alex6323) [reply bee-network/bee-gossip/src/network/host.rs]: Reduced this PR. This will go into the bee-gossip rewrite.

File: bee-network/bee-autopeering/src/peer/peer_id.rs (hunk 7 above)
Comment (Alex6323) [reply bee-network/bee-autopeering/src/peer/peer_id.rs:123]: No, if that would panic, then something with the node's identity crypto is wrong. Maybe somone messed with the identity.key file or some other serious issue. The node is definitely not in sane state to even start.

File: bee-network/bee-autopeering/src/local/mod.rs (hunk 6 above)
Comment (Alex6323) [reply bee-network/bee-autopeering/src/local/mod.rs]: no, but I removed it that function as it was redundant.

File: bee-identity/src/identity.rs [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    pub fn public_key(&self) -> Ed25519PublicKey {
+        self.keypair.public()
+    }
+
+    /// Returns the corresponding [`libp2p_core::PeerId`].
+    pub fn peer_id(&self) -> PeerId {
+        PeerId::from_public_key(&PublicKey::Ed25519(self.keypair.public()))
+    }
+
+    /// Returns whether this identity has been newly generated.
```
Comment (samuel-rufi) [inline bee-identity/src/identity.rs]: what does "fresh" mean in this context? also when isn't it considered fresh anymore?  :trollface: Maybe elaborate in the comment

**Review by samuel-rufi — CHANGES_REQUESTED** [review]

File: bee-identity/src/identity.rs (hunk 17 above)
Comment (Alex6323) [reply bee-identity/src/identity.rs]: The doc you marked is literally my definition of fresh in this context. Basically the same as `new`, but I got a bit bored of that term. :smile:

Comment (Alex6323) [reply bee-identity/src/identity.rs]: For example, if you restore an identity then it existed before, right, so it's not fresh anymore. Want me to rename to `is_new`?

File: bee-identity/src/identity.rs [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+#[derive(Clone, Debug)]
+pub struct Identity {
+    keypair: Ed25519Keypair,
+    is_fresh: bool,
+}
+
+impl Identity {
+    /// Creates a new identity.
+    pub fn new() -> Self {
+        Self::default()
```
Comment (samuel-rufi) [inline bee-identity/src/identity.rs]: isn't in this case ` is_fresh: bool` defaulting to `false` but it should be `true`?

File: bee-identity/src/identity.rs (hunk 17 above)
Comment (samuel-rufi) [reply bee-identity/src/identity.rs]: Oh understand 😄 no strong opionion on this, so as you desire. i think i would just change the comment then to something like: "Returns whether this identity has been newly generated or restored from the file system."

File: bee-identity/src/identity.rs (hunk 18 above)
Comment (Alex6323) [reply bee-identity/src/identity.rs]: have a look at the `default` implementation of `Identity` :trollface:

**Review by samuel-rufi — DISMISSED** [review]

File: bee-identity/src/identity.rs (hunk 17 above)
Comment (grtlr) [reply bee-identity/src/identity.rs]: Maybe `restored_from_file` is more appropriate then?

File: bee-identity/src/pem_file.rs [hunk 19]
```diff
@@ -1,20 +1,28 @@
...
-use bee_gossip::{Keypair, SecretKey};
 use ed25519::KeypairBytes;
 use pkcs8::{DecodePrivateKey, EncodePrivateKey, LineEnding};
 
+use crate::ed25519::{Keypair, SecretKey};
+
+/// PEM file errors.
 #[derive(Debug, thiserror::Error)]
 pub enum PemFileError {
+    /// Read error.
```
Comment (grtlr) [inline bee-identity/src/pem_file.rs]: nitpick: I'd put `#[allow(missing_docs)]` on `PemFileError`.

**Review by grtlr — DISMISSED** [review]

File: bee-identity/src/identity.rs (hunk 17 above)
Comment (samuel-rufi) [reply bee-identity/src/identity.rs]: good idea, maybe `is_restored_from_file` ? would be more clear that it will return a bool

File: bee-identity/src/pem_file.rs (hunk 19 above)
Comment (Alex6323) [reply bee-identity/src/pem_file.rs]: Done

File: bee-identity/src/identity.rs (hunk 17 above)
Comment (Alex6323) [reply bee-identity/src/identity.rs]: changed it to `is_restored`. I think as long as we have no other source restoring from it's simpler and clear already.

**Review by grtlr — DISMISSED** [review]

**Review by samuel-rufi — DISMISSED** [review]

File: bee-identity/CHANGELOG.md [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+<!-- ## Unreleased - YYYY-MM-DD
+
+### Added
+
+- New `./dashboard` route;
+- Automatically migrate identity from `Config.toml`;
+- Generate new identity if none is present;
```
Comment (thibault-martinez) [inline bee-identity/CHANGELOG.md]: Remove

File: bee-identity/CHANGELOG.md [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+### Changed
+
+- P2P identity is now stored in a `identity.key` file;
```
Comment (thibault-martinez) [inline bee-identity/CHANGELOG.md]: Remove

File: bee-identity/CHANGELOG.md [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+
+### Deprecated
+
+### Removed
+
+- `p2p_identity` tool;
```
Comment (thibault-martinez) [inline bee-identity/CHANGELOG.md]: Remove

File: bee-identity/CHANGELOG.md [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+### Fixed
+
+### Security -->
+
+## 0.1.0 - 2021-04-xx
```
Comment (thibault-martinez) [inline bee-identity/CHANGELOG.md]: Today's date

File: bee-identity/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+pub mod ed25519 {
+    #[doc(inline)]
+    pub use libp2p_core::identity::ed25519::{Keypair, PublicKey, SecretKey};
+}
+pub mod pem_file;
+
+#[doc(inline)]
+pub use libp2p_core::PeerId;
+
+pub use crate::{
```
Comment (thibault-martinez) [inline bee-identity/src/lib.rs]:
```suggestion
pub use self::{
```

File: bee-identity/CHANGELOG.md (hunk 20 above)
Comment (Alex6323) [reply bee-identity/CHANGELOG.md]: Done

File: bee-identity/CHANGELOG.md (hunk 21 above)
Comment (Alex6323) [reply bee-identity/CHANGELOG.md]: Done

File: bee-identity/CHANGELOG.md (hunk 22 above)
Comment (Alex6323) [reply bee-identity/CHANGELOG.md]: You really mean it :trollface: Done!

File: bee-identity/CHANGELOG.md (hunk 23 above)
Comment (Alex6323) [reply bee-identity/CHANGELOG.md]: Done

File: bee-identity/src/lib.rs (hunk 24 above)
Comment (Alex6323) [reply bee-identity/src/lib.rs]: sorry...overlooked that I could have committed your suggestion. I hope you don't mind :grimacing:

File: bee-network/bee-autopeering/src/peer/peer_id.rs [hunk 25]
```diff
@@ -115,6 +116,14 @@ impl AsRef<[u8]> for PeerId {
...
```
Comment (thibault-martinez) [inline bee-network/bee-autopeering/src/peer/peer_id.rs:123]: Put a PANIC comment please

File: bee-node/bee-node/src/fullnode/builder.rs [hunk 26]
```diff
@@ -207,14 +208,10 @@ fn add_node_resources<S: NodeStorageBackend>(builder: FullNodeBuilder<S>) -> Res
     // TODO block ? Make new async ?
     let storage = S::start(storage_cfg).map_err(|e| CoreError::StorageBackend(Box::new(e)))?;
 
-    if config.local().is_new() {
+    if !config.identity().is_restored() {
         storage
             .set_health(StorageHealth::Healthy)
             .map_err(|e| CoreError::StorageBackend(Box::new(e)))?;
-
-        return Err(FullNodeError::InvalidOrNoIdentityPrivateKey(
```
Comment (thibault-martinez) [inline bee-node/bee-node/src/fullnode/builder.rs:222]: Why is this removed ?

File: bee-network/bee-autopeering/src/peer/peer_id.rs (hunk 25 above)
Comment (Alex6323) [reply bee-network/bee-autopeering/src/peer/peer_id.rs:123]: Done

File: bee-node/bee-node/src/fullnode/builder.rs (hunk 26 above)
Comment (Alex6323) [reply bee-node/bee-node/src/fullnode/builder.rs:222]: Because ... if the identity hasn't been restored ... then a new valid one was created.

Comment (Alex6323) [reply bee-node/bee-node/src/fullnode/builder.rs:222]: Actually I am not sure I still understand the logic behind that if-branch. Why are we setting the storage to healthy and then exit?

Comment (Alex6323) [reply bee-node/bee-node/src/fullnode/builder.rs:222]: Clarified that the whole if-branch can be removed.

File: bee-identity/src/identity.rs [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /// Restores an [`Identity`] from a [`Keypair`](crate::ed25519::Keypair).
+    pub fn from_keypair(keypair: Ed25519Keypair) -> Self {
+        Self {
+            keypair,
+            restored_from_file: true,
```
Comment (thibault-martinez) [inline bee-identity/src/identity.rs]: True ?

Comment (thibault-martinez) [reply bee-identity/src/identity.rs]: It's a bit weird to me that this is called `restored_from_file` when obviously it can be restored from pretty much anything

Comment (Alex6323) [reply bee-identity/src/identity.rs]: you're right. `restored` is enough.

Comment (Alex6323) [reply bee-identity/src/identity.rs]: Changed.

File: bee-identity/src/identity.rs [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+            Err(e) => {
+                log::error!("Could not extract private key from PEM file: {}", e);
+                Err(Box::new(e))
+            }
+        }
+    }
+}
+
+impl Default for Identity {
+    fn default() -> Self {
```
Comment (thibault-martinez) [inline bee-identity/src/identity.rs]: I think we once talked about that and preferred to have a `generate` method instead of a Default impl

Comment (Alex6323) [reply bee-identity/src/identity.rs]: true!

Comment (Alex6323) [reply bee-identity/src/identity.rs]: Fixed.

File: bee-identity/src/identity.rs [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+    pub fn generate() -> Self {
+        let keypair = Ed25519Keypair::generate();
+
+        Self {
+            keypair,
+            restored: false,
+        }
+    }
+
+    /// Restores an [`Identity`] from a [`Keypair`](crate::ed25519::Keypair).
```
Comment (thibault-martinez) [inline bee-identity/src/identity.rs]: From an Ed25519Keypair

File: bee-identity/src/identity.rs [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+    pub fn from_keypair(keypair: Ed25519Keypair) -> Self {
+        Self {
+            keypair,
+            restored: true,
+        }
+    }
+
+    /// Returns the backing [`Keypair`](crate::ed25519::Keypair) of this identity.
```
Comment (thibault-martinez) [inline bee-identity/src/identity.rs]: Same

File: bee-identity/src/identity.rs [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+    pub fn keypair(&self) -> &Ed25519Keypair {
+        &self.keypair
+    }
+
+    /// Returns the [`PublicKey`](crate::ed25519::PublicKey) of this identity.
```
Comment (thibault-martinez) [inline bee-identity/src/identity.rs]: Same

File: bee-identity/src/identity.rs [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+    pub fn public_key(&self) -> Ed25519PublicKey {
+        self.keypair.public()
+    }
+
+    /// Returns the corresponding [`libp2p_core::PeerId`].
+    pub fn peer_id(&self) -> PeerId {
+        PeerId::from_public_key(&PublicKey::Ed25519(self.keypair.public()))
+    }
+
+    /// Returns whether this identity has been restored from a file.
```
Comment (thibault-martinez) [inline bee-identity/src/identity.rs]:
```suggestion
    /// Returns whether this identity has been restored from an already existing keypair.
```

**Review by thibault-martinez — APPROVED** [review]

**Review by samuel-rufi — APPROVED** [review]
